lib/elixir/lib/calendar/datetime.ex

```diff
@@ -585,8 +593,12 @@ defmodule DateTime do
 
   """
   @spec convert(Calendar.datetime, Calendar.calendar) :: {:ok, t} | {:error, :incompatible_calendars}
-  def convert(%DateTime{calendar: calendar} = datetime, calendar) do
-    {:ok, datetime}
+  def convert(%{calendar: calendar, year: year, month: month, day: day,
+                hour: hour, minute: minute, second: second, microsecond: microsecond,
+                time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset}, calendar) do
+    {:ok, %DateTime{calendar: calendar, year: year, month: month, day: day,
+                    hour: hour, minute: minute, second: second, microsecond: microsecond,
+                    time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset}}
```

Comment:
I agree copy-pasting is ugly. However, I suppose someone could have a `MyDateTime` struct with all datetime fields (thus satisfying `Calendar.datetime` type) plus another `foo` field and then `struct!(DateTime, fields)` would break. Not sure it'd be a problem in practice.